module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,529 @@
+        }
+        $request_token = $this->getRequestToken($code);
+        $state = $request->getQuery()->get('state');
+        $stateIsValid = $state == $this->session->oidc_state;
+        unset($this->session->oidc_state);
+        if (!$stateIsValid) {
+            throw new AuthException('authentication_error_admin: bad state');
+        }
+
+        $claims = $this->decodeJWT($request_token->id_token);
+
+        if (!$this->validateIssuer($claims->iss)) {
+            throw new AuthException('authentication_error_admin: wrong issuer');
+        }
+        $claimsValid = $this->verifyJwtClaims($claims);
+        unset($this->session->oidc_nonce);
+        if (!$claimsValid) {
+            throw new AuthException('authentication_error: not valid claims');
+        }
+
+        $accessToken = $request_token->access_token;
+        $userInfo = $this->getUserInfo($accessToken);
+        return $this->setUserAttributes($userInfo);
+    }
+
+    /**
+     * Set user attributes from user info claim
+     *
+     * @param object $userInfo User info claim object
+     *
+     * @return UserEntityInterface
+     * @throws AuthException
+     * @throws PasswordSecurityException
+     * @throws \Exception
+     */
+    protected function setUserAttributes(object $userInfo): UserEntityInterface
+    {
+        $availableAttributes = [
+            'firstname',
+            'lastname',
+            'email',
+            'cat_id',
+            'cat_username',
+            'cat_password',
+            'college',
+            'major',
+            'home_library',
+        ];
+        $userService = $this->getUserService();
+        $user = $this->getOrCreateUserByUsername(($this->getConfig('username_prefix') ?? '') . $userInfo->sub);
+        $attrMappings = array_filter(
+            $this->getAttributesMappings(),
+            function ($key) use ($availableAttributes) {
+                return in_array($key, $availableAttributes);
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+        $catPassword = null;
+        foreach ($attrMappings as $userAttr => $infoAttr) {
+            $attrValue = $this->getAttributeValue($userInfo, $infoAttr);
+            if (!empty($attrValue)) {
+                if ($userAttr === 'email') {
+                    $userService->updateUserEmail($user, $attrValue);
+                    continue;
+                }
+                if ($userAttr === 'cat_password') {
+                    $catPassword = $attrValue;
+                    continue;
+                }
+                $this->setUserValueByField($user, $userAttr, $attrValue);
+            }
+        }
+        $catUsername = $user->getCatUsername();
+        if (!empty($catUsername)) {
+            $currentPassword = $this->ilsAuthenticator->getCatPasswordForUser($user);
+            $this->ilsAuthenticator->setUserCatalogCredentials($user, $catUsername, $catPassword ?? $currentPassword);
```

Comment:
I think you might want to refactor this to:

```suggestion
            $finalCatPassword = $catPassword ?? $this->ilsAuthenticator->getCatPasswordForUser($user);
            $this->ilsAuthenticator->setUserCatalogCredentials($user, $catUsername, $finalCatPassword);
```

That way we can avoid doing an expensive ILS lookup when we don't need to.

Reply:
And it is a bit more readable I think, thank you. Done.